packages/api/src/User/User.service.ts

```diff
@@ -36,4 +38,62 @@ export class UserService {
     user.profileImage = `https://www.gravatar.com/avatar/${md5(input.email)}`;
     return this.userRepository.save(user);
   }
+
+  /**
+   * retreive and or reset streak for not completeing modules
+   * @param user
+   */
+  async retreiveStreak(user: UserWithPassword) {
+    const now = new Date();
+
+    const lastCompleted = await this.userModuleService.findLatestByUser(
+      user.id
+    );
+
+    if (!lastCompleted) return user.streak;
+
+    const diffInHours = Math.floor(
+      (now.valueOf() - lastCompleted.valueOf()) / 1000 / 60 / 60
+    );
+
+    if (diffInHours >= 24) {
+      await this.userRepository.update({ id: user.id }, { streak: 0 });
+      return 0;
+    }
+    return user.streak;
+  }
+
+  /**
+   * Retreive and or extend streak of completing modules
+   * @param user
+   */
+  async incrementStreak(userOrId: string | UserWithPassword) {
+    const user = typeof userOrId === 'string'
+      ? await this.userRepository.findOneOrFail({ id: userOrId })
+      : userOrId;
+
+    const { streak } = user;
+
+    const userModules = await this.userModuleService.findByUser(user.id);
+
+    if (!userModules) return user.streak;
+
+    // if there are more than one modules completed
+    if (userModules.length > 1) {
```

Comment:
This appears to only check if the first two UserModules returned for this user were completed at the same date, and if not same date then increment streak.

I'm not sure if this handles all of the use cases.
E.g. if user completed one module each day for past 5 days in a row, how would this code work?
(I would expect it to increment the streak from 5 to 6)

Reply:
Hey @shaneSheridan, this is what I'm thinking:  if user completes a module on day 1, if it is the first module completed that day, we would increment the streak. If user would then complete a second module in that same day, because one had been completed earlier we would not increment the streak. This would continue every day or until the streak is broken.

The incrementStreak method checks to see how many modules have been completed and, more importantly, if two or more were completed within the same day. We will always increment the streak if there is only one module completed. If there are two or more modules, we will check to see if the first two (modules we get back for a user are always be sorted by the completedAt date in descending order) were completed within the same day. If yes, we do not increment. If no, we increment.

The retreiveStreak method determines whether or not the streak needs to be reset. The incrementStreak method determines whether or not the streak needs to be incremented.